File: implementation/CustomStack.java
/*
 * Implementation of stack
 * push(x) to insert x into stack
 * pop() to return top most element
 * peek() to print top most element
 * isEmpty to check if stack is empty or not
 * size() to print the size of the stack
 * */

class Node {
    int value;
    Node next;

    public Node(int value) {
        this.value = value;
    }
}

public class CustomStack{

    Node head = null;
    int stack_size = 0;

    public void push(int value) {

        Node node = new Node(value);

        node.next = head;

        head = node;

        stack_size++;
    }

    public int pop() {

        if (stack_size > 0) {

            int value = head.value;

            head = head.next;
            return value;

        } else {
            return -1;
        }
    }

    public int peek() {
        if (head != null) {
            return head.value;
        } else{
            return -1;
        }
    }


    public int size() {
        return stack_size;
    }

    public boolean isEmpty() {
        return head == null;
    }

    public static void main(String[] args) {

        CustomStack stack = new CustomStack();

        assert stack.isEmpty() == true;

        stack.push(4);

        stack.push(3);

        assert stack.peek() == 3;

        assert stack.size() == 2;

        stack.push(2);

        assert stack.pop() == 2;

        assert stack.pop() == 3;

        assert stack.pop() == 4;

    }
}

File: adt/CustomStack.py
"""
  Implementation of stack
  push(x) to insert x into stack
  pop() to return top most element
  peek() to print top most element
  isEmpty to check if stack is empty or not
  size() to print the size of the stack
 **/
"""
class Node(object):
    def __init__(self, value):
        self.value = value
        self.next = None

class CustomStack(object):
    def __init__(self):
        self.head = None
        self.size = 0

    def push(self, value):

        curr = Node(value)
        curr.next = self.head
        self.head = curr
        self.size += 1

    def pop(self):
        ret = None
        if self.head != None:
            ret = self.head.value
            self.head = self.head.next
            self.size-= 1

        return ret

    def peek(self):
        ret = None
        if self.head != None:
            ret =  self.head.value

        return ret

    def size(self):
        return self.size

if __name__=="__main__":
    cs = CustomStack()

    cs.push(4)
    cs.push(3)

    assert cs.peek() == 3

    cs.push(2)

    assert cs.pop() == 2

    assert cs.pop() == 3

    assert cs.pop() == 4
